src/Data/Text/Utf8.hs

```diff
@@ -326,108 +260,50 @@ unsafeIndexCodeUnit (Text !u8data !off !_len) (CodeUnitIndex !index) =
+  TextUnsafe.takeWord8 len $ TextUnsafe.dropWord8 begin text
 
--- $generalFunctions
+-- $functionsOnArrays
 --
--- These functions are available in @text@ as well and should be removed once this library moves to @text-2@.
--- You should be able to use these by doing @import qualified Data.Text.Utf8 as Text@ just like you would with @text@.
---
--- NOTE: The 'Text' instances for @Show@, @Eq@, @Ord@, @IsString@, @FromJSON@, @ToJSON@ and @Hashable@ in this file also fall in this category.
-
--- | TODO: Inefficient placeholder implementation.
-concat :: [Text] -> Text
-concat = pack . concatMap unpack
+-- Functions for working with 'TextArray.Array' values.
 
--- | See 'Data.Text.dropWhile'.
-dropWhile :: (Char -> Bool) -> Text -> Text
-dropWhile predicate text =
-  let
-    len = codeUnitIndex (lengthUtf8 text)
-    go i
-      | i >= CodeUnitIndex len = i
-      | otherwise =
-        let
-          (codeUnits, codePoint) = unsafeIndexCodePoint text i
-        in
-          if predicate codePoint then
-            go $ i + codeUnits
-          else
-            i
-
-    prefixEnd = go 0
-  in
-    unsafeSliceUtf8 prefixEnd (CodeUnitIndex len - prefixEnd) text
+-- | See 'Data.Primitive.isByteArrayPinned'.
+isArrayPinned :: TextArray.Array -> Bool
+isArrayPinned (TextArray.ByteArray ba#) = Exts.isTrue# (Exts.isByteArrayPinned# ba#)
 
--- | Checks whether a text is the empty string.
-null :: Text -> Bool
-null (Text _ _ len) = len == 0
+-- | See 'Data.Primitive.byteArrayContents'.
+arrayContents :: TextArray.Array -> Exts.Ptr Word8
+arrayContents (TextArray.ByteArray ba#) = Exts.Ptr (Exts.byteArrayContents# ba#)
 
--- | TODO: Inefficient placeholder implementation.
-pack :: String -> Text
-pack = go . stringToByteArray
-  where
-    go !arr = Text arr 0 $ sizeofByteArray arr
-
--- | TODO: Inefficient placeholder implementation.
--- See 'Data.Text.replicate'
-replicate :: Int -> Text -> Text
-replicate n = pack . Prelude.concat . Prelude.replicate n . unpack
-
--- | TODO: Inefficient placeholder implementation.
--- This function implements very basic string search. It's @text@ counterpart is 'Data.Text.Internal.Search.indices', which implements the Boyer-Moore algorithm.
--- Since we have this function only to check whether our own Boyer-Moore implementation works, it would not make much sense to implement it using the same algorithm.
--- Once we can use @text-2@, we can compare our implementation to the official @text@ one which presumably works.
-indices :: Text -> Text -> [Int]
-indices needle haystack
-  | needleLen == 0 = []
-  | otherwise = go 0 0
+-- | Decode a code point at the given 'CodeUnitIndex'.
+-- Returns garbage if there is no valid code point at that position.
+-- Does not perform bounds checking.
+-- See 'decode2', 'decode3' and 'decode4' for the expected format of multi-byte code points.
+{-# INLINE unsafeIndexCodePoint' #-}
+unsafeIndexCodePoint' :: TextArray.Array -> CodeUnitIndex -> (CodeUnitIndex, CodePoint)
```

Comment:
It is likely faster to use [`iterArray`](https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Unsafe.html#v:iterArray).

Reply:
First of all, thank you for your review! Hope you liked the article.

We compared against `iterArray` in an earlier iteration and came to the conclusion that it is slower. Looking into its source code, this is probably because it uses [`unsafeIndex`](https://hackage.haskell.org/package/text-2.0/docs/src/Data.Text.Array.html#unsafeIndex) which performs bounds checks on the `ByteArray`. Core-2-Core optimizations will likely turn this into a single bounds check, however, `unsafeIndexCodePoint'` performs none. There was no noticeable difference in performance between using `<` and bitwise operations to check the MSBs of the first code unit.

To make sure that we aren't missing out, I just ran our benchmark against three versions of the UTF-8 code:

```
$ git log --no-decorate --oneline | head -n 3
78fc4bc Replace unsafeIndexCodePoint' by iterArray
0ab379e Remove redundant remaining parameter
22dff79 Re-add isCaseInvariant function
```

- 22dff79 is the current state of this PR
- 0ab379e removes the `remaining` parameter from the functions in the inner loop which is remnant of the UTF-16 version.
- 78fc4bc uses `iterArray` instead of our own function

The results are:

```
haskell-22dff79.stats:
  mean time: 2.726 ± 0.027 seconds
   min time: 2.618 seconds

haskell-0ab379e.stats:
  mean time: 2.505 ± 0.013 seconds
   min time: 2.417 seconds

haskell-78fc4bc.stats:
  mean time: 3.261 ± 0.011 seconds
   min time: 3.194 seconds
```

As you can see, `iterArray` is quite a bit slower. However, thanks to your review, I removed the `remaining` parameter which gave us another small improvement! I will cherry-pick that commit into this PR :rocket: 

I also took a superficial look at the generated Core, suspecting that the conversion to and from `CodeUnitIndex` could slow the `iterArray` version down. That's not the case though; it seems like the bounds checks are the problem indeed.

Reply:
I took another look at the Core and, it seems like the main difference lies in how the length of the code unit sequence constituting a code point is determined:

For `unsafeIndexCodePoint'`, the ASCII case looks like this:

```haskell
-- ipv4_sjGM is the array
-- ww2_sjHC is our current index
case indexWord8Array# ipv4_sjGM ww2_sjHC of r#1_sjHG
{ __DEFAULT ->
case ltWord# r#1_sjHG 192## of {
  __DEFAULT -> -- Check other cases
  1# -> -- Code point consists of a single code unit, use that
}
```

For `iterArray`, it looks like this:

```haskell
-- ipv4_sGDw is the array
-- ww2_sGEp is our current index
case indexWord8Array# ipv4_sGDw ww2_sGEp of r#1_sGEt
{ __DEFAULT ->
case xor# r#1_sGEt 255## of sat_sGEv { __DEFAULT ->
case clz8# sat_sGEv of sat_sGEw { __DEFAULT ->
case word2Int# sat_sGEw of c#1_sGEu { __DEFAULT ->
case <=# c#1_sGEu 0# of sat_sGEx { __DEFAULT ->
case xorI# c#1_sGEu sat_sGEx of ds_sGEy {
  __DEFAULT -> -- 4 code units
  1# -> -- 1 code unit
  2# -> -- 2 code units
  3# -> -- 3 code unit
}
```

My assumption is that since we mostly handle ASCII data, the additional operations for each character cause the overhead. For CJK or non-BMP texts `iterArray` would probably be faster since it uses less jumps. Perhaps we will try to measure that in another micro-benchmark, e.g. ASCII vs. CJK vs. non-BMP vs. random strings.

Reply:
@poorlyknitsweater Could you please try replacing all occurrences of `chr` with https://hackage.haskell.org/package/text-2.0/docs/Data-Text-Internal-Unsafe-Char.html#v:unsafeChr8 and similar? The reason is that `chr` involves a redundant safety check: https://hackage.haskell.org/package/base-4.16.1.0/docs/src/GHC-Char.html#chr

Reply:
Thanks again! I tried this out, but the benchmark doesn't show a significant difference :(

I compared 952e432 (from yesterday) and 7b8be51, which replaces `chr` by either `unsafeChr8` or `unsafeChr`:

```
haskell-952e432.stats:
  mean time: 2.792 ± 0.016 seconds
   min time: 2.681 seconds

haskell-7b8be51.stats:
  mean time: 2.774 ± 0.015 seconds
   min time: 2.675 seconds
```

Both Core and CMM outputs show that the jump was actually eliminated:

952e432:

```haskell
case ltWord# r#1_s11cE 192## of {
  __DEFAULT -> -- Not ASCII
  1# -> -- ASCII
    case word2Int# r#1_s11cE of i#_s11eg { __DEFAULT -> -- These two lines are not present anymore in -ddump-opt-cmm
    case int2Word# i#_s11eg of sat_s11eh { __DEFAULT ->
    case leWord# sat_s11eh 1114111## of { -- Unnecessary check introduced by chr
      __DEFAULT -> ($wlvl i#_s11eg) `cast` <Co:4>; -- error that's never thrown
      1# ->
        case chr# i#_s11eg of ww4_s11ej { __DEFAULT -> -- Convert to ww4_s11ej :: Char#
        case ww3_s11cC of wild4_s11ek { -- Switch on current aho-corasick state
```

7b8be51:

```haskell
case ltWord# r#1_s10r6 192## of {
  __DEFAULT -> -- Not ASCII
  1# -> -- ASCII
    case word2Int# r#1_s10r6 of sat_s10sD { __DEFAULT -> -- Convert to Int# (Haskell source uses unsafeChr8 here)
    case chr# sat_s10sD of ww4_s10sC { __DEFAULT -> -- Convert to Char#
    case ww3_s10r4 of wild4_s10sE { -- Switch on current aho-corasick state
```

So it's kind of weird that this doesn't improve the benchmark times, my assumption is that this branch is just very predictable since it's never taken.

Reply:
@Bodigrim we will merge this PR now, if you have any further ideas for improvement, please create an issue! Thank you again for the review :)